examples/omnivore/data/datasets.py

```diff
@@ -0,0 +1,242 @@
+        img_rgb = PIL.Image.open(rgb_path)
+        tensor_rgb = T.ToTensor()(img_rgb)
+
+        tensor_d = self._get_disparity_tensor(image_dir)
+
+        tensor_rgbd = torch.cat((tensor_rgb, tensor_d), dim=0)
+        return tensor_rgbd
+
+    def _get_sunrgbd_scene_class(self, image_dir):
+        with open(os.path.join(image_dir, "scene.txt"), "r") as fin:
+            scene_class = fin.read().strip()
+        return scene_class
+
+    def __getitem__(self, idx):
+        # return tuple of image (H W C==4) and scene class index
+        image_dir = self.image_dirs[idx]
+        x_rgbd = self._read_sunrgbd_image(image_dir)
+        scene_class = self._get_sunrgbd_scene_class(image_dir)
+        scene_idx = self.class_to_idx[scene_class]
+
+        if self.transform:
+            x_rgbd = self.transform(x_rgbd)
+
+        if self.target_transform:
+            scene_idx = self.target_transform(scene_idx)
+
+        return x_rgbd, scene_idx
+
+
+class ConcatIterable:
+    """
+    ConcatIterable is used to group iterable object.
+    When user iterate on this object, we will sample random iterable and return their
+    item with coresponding output_key.
+    With repeat_factors, user can do upsampling or downsampling to the iterables.
+    We mainly use this class to concat different data loader during training.
+
+    Args:
+        iterables: the iterable objects that will be grouped
+        output_keys: List of keys that is used to identify the iterable output.
+            The list length should be the same as number of iterables.
+        repeat_factors: List of numbers that represent the upsampling / downsampling factor
+            to the coresponding iterables. Should have same length as iterables.
+        seed: the seed for randomness
+    """
+
+    def __init__(self, iterables, output_keys, repeat_factors, seed=42):
+        self.iterables = iterables
+        self.output_keys = output_keys
+        self.repeat_factors = repeat_factors
+        self.seed = seed
+        self.num_iterables = len(self.iterables)
+        assert self.num_iterables == len(output_keys)
+        assert self.num_iterables == len(repeat_factors)
+
+        # The iterator len is adjusted with repeat_factors
+        self.iterator_lens = [
+            int(repeat_factors[i] * len(itb)) for i, itb in enumerate(self.iterables)
+        ]
+        self.max_total_steps = sum(self.iterator_lens)
+        self.indices = None
+        self.iterators = None
+
+        # self.step_counter == None indicate that self.indices are not yet initialized
+        self.step_counter = None
+
+    def init_indices(self, epoch=0, shuffle=False):
+        # We should initiate indices for each epoch, especially if we want to shuffle
+        self.step_counter = 0
+
+        self.iterators = [iter(dl) for dl in self.iterables]
+        self.indices = torch.cat(
+            [
+                torch.ones(self.iterator_lens[i], dtype=torch.int32) * i
```

Comment:
Are the indices the same within each iterable (I assume an iterable corresponds to one dataset/modality)?
E.g., `0, 0, 0, ..., 0` for dataset1, `1, 1, ,1..., 1` for dataset2, etc. ?

Reply:
yes, correct. This indices will later be used to determine which `dataset` that we will return from.